File: pizza/src/main/java/SolverStack.java
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class SolverStack {

    private static class SolveStack {
        final int pizzaIndex;
        final List<Pizza> candidates;
        final int sumCandidates;

        private SolveStack(int pizzaIndex, List<Pizza> candidates, int sumCandidates) {
            this.pizzaIndex = pizzaIndex;
            this.candidates = candidates;
            this.sumCandidates = sumCandidates;
        }
    }

    private int deltaMin;
    private List<Pizza> candidateSolution = new ArrayList<>();
    private List<SolveStack> stack = new ArrayList<>();

    public Solution solve(File fileIn) throws IOException {
        ProblemParser parser = new ProblemParser();
        Problem problem = parser.parse(fileIn);
        return solve(problem);
    }

    public Solution solve(Problem problem) {
        deltaMin = Integer.MAX_VALUE;
        candidateSolution.clear();
        stack.clear();

        Solution solution = new Solution(problem);
        List<Pizza> pizzas = problem.pizzas;
        if (pizzas.isEmpty()) {
            return solution;
        }

        List<Pizza> candidates = new ArrayList<>();
        SolveStack stackElement = new SolveStack(pizzas.size() - 1, candidates, 0);
        stack.add(stackElement);
        List<Pizza> solved = solve(problem.M, pizzas);
        if (solved != null) {
            solution.pizzas.addAll(solved);
        } else {
            solution.pizzas.addAll(candidateSolution);
        }
        return solution;
    }

    private List<Pizza> solve(final int maximumSlices, final List<Pizza> pizzas) {
        SolveStack stackElement;
        int pizzaIndex;
        List<Pizza> candidates;
        int sumCandidates;
        int gcCounter = 0;

        while (stack.size() > 0) {
            stackElement = stack.remove(0);
            pizzaIndex = stackElement.pizzaIndex;
            candidates = stackElement.candidates;
            sumCandidates = stackElement.sumCandidates;

            if (pizzaIndex >= 0) {
                if (candidates.isEmpty()) {
                    System.out.println("solve: [] " + sumCandidates);
                } else {
                    System.out.println("solve: [" + candidates.get(0) + ", ...] " + sumCandidates);
                }

                Pizza candidate = pizzas.get(pizzaIndex);

                List<Pizza> pick = new ArrayList<>(candidates);
                pick.add(0, candidate);
                int sumPick = sumCandidates + candidate.slices;
                int delta = maximumSlices - sumPick;
                if (delta == 0) {
                    return pick;
                }

                List<Pizza> skip = candidates;
                stackElement = new SolveStack(pizzaIndex - 1, skip, sumCandidates);
                stack.add(0, stackElement);

                if (delta > 0) {
                    if (delta < deltaMin) {
                        deltaMin = delta;
                        candidateSolution = pick;
                    }
                    stackElement = new SolveStack(pizzaIndex - 1, pick, sumPick);
                    stack.add(0, stackElement);
                }
            }

            gcCounter++;
            if (gcCounter >= 100) {
                gcCounter = 0;
                System.gc();
            }
        }

        return null;
    }
}

File: pizza/src/main/java/SolverStack.kt
import java.io.File
import java.io.IOException
import java.util.*

class SolverStack {
    private class SolveStack(val pizzaIndex: Int, val candidates: List<Pizza>, val sumCandidates: Int)

    private var deltaMin = 0
    private var candidateSolution: MutableList<Pizza> = ArrayList()
    private val stack: MutableList<SolveStack> = ArrayList()

    @Throws(IOException::class)
    fun solve(fileIn: File?): Solution {
        val parser = ProblemParser()
        val problem = parser.parse(fileIn!!)
        return solve(problem)
    }

    fun solve(problem: Problem): Solution {
        deltaMin = Int.MAX_VALUE
        candidateSolution.clear()
        stack.clear()
        val solution = Solution(problem)
        val pizzas: List<Pizza> = problem.pizzas
        if (pizzas.isEmpty()) {
            return solution
        }
        val candidates: List<Pizza> = ArrayList()
        val stackElement = SolveStack(pizzas.size - 1, candidates, 0)
        stack.add(stackElement)
        val solved = solve(problem.maximumSlicesCount, pizzas)
        if (solved != null) {
            solution.pizzas.addAll(solved)
        } else {
            solution.pizzas.addAll(candidateSolution)
        }
        return solution
    }

    private fun solve(maximumSlices: Int, pizzas: List<Pizza>): List<Pizza>? {
        var stackElement: SolveStack
        var pizzaIndex: Int
        var candidates: List<Pizza>
        var sumCandidates: Int
        var gcCounter = 0
        while (stack.size > 0) {
            stackElement = stack.removeAt(0)
            pizzaIndex = stackElement.pizzaIndex
            candidates = stackElement.candidates
            sumCandidates = stackElement.sumCandidates
            if (pizzaIndex >= 0) {
                if (candidates.isEmpty()) {
                    println("solve: [] $sumCandidates")
                } else {
                    println("solve: [" + candidates[0] + ", ...] " + sumCandidates)
                }
                val candidate = pizzas[pizzaIndex]
                val pick: MutableList<Pizza> = ArrayList(candidates)
                pick.add(0, candidate)
                val sumPick = sumCandidates + candidate.slices
                val delta = maximumSlices - sumPick
                if (delta == 0) {
                    return pick
                }
                val skip = candidates
                stackElement = SolveStack(pizzaIndex - 1, skip, sumCandidates)
                stack.add(0, stackElement)
                if (delta > 0) {
                    if (delta < deltaMin) {
                        deltaMin = delta
                        candidateSolution = pick
                    }
                    stackElement = SolveStack(pizzaIndex - 1, pick, sumPick)
                    stack.add(0, stackElement)
                }
            }
            gcCounter++
            if (gcCounter >= 100) {
                gcCounter = 0
                System.gc()
            }
        }
        return null
    }
}